src/redux/slices/search.js

```diff
@@ -0,0 +1,80 @@
+import { createSlice } from '@reduxjs/toolkit';
+
+export const searchSlice = createSlice({
+  name: 'search',
+  initialState: {
+    startAddress: '',
+    endAddress: '',
+    startLocation: null,
+    endLocation: null,
+    distance: null,
+    directions: null,
+    elevationProfile: null,
+    path: null,
+  },
+  reducers: {
+    setStartAddress: (state, action) => {
+      state.startAddress = action.payload;
+    },
+    setEndAddress: (state, action) => {
+      state.endAddress = action.payload;
+    },
+    setStartLocation: (state, action) => {
+      state.startLocation = action.payload;
+    },
+    setEndLocation: (state, action) => {
+      state.endLocation = action.payload;
+    },
+    setDistance: (state, action) => {
+      state.distance = action.payload;
+    },
+    setDirections: (state, action) => {
+      state.directions = action.payload;
+    },
+    setElevationProfile: (state, action) => {
+      state.elevationProfile = action.payload;
+    },
+    setPath: (state, action) => {
+      state.path = action.payload;
+    },
+    clearPath: (state) => ({
+      ...state,
+      path: null,
+      directions: null,
+      elevationProfile: null,
+    }),
+    clearMarkers: (state) => ({
+      ...state,
+      startLocation: null,
+      endLocation: null,
+      startAddress: '',
+      endAddress: '',
+    }),
+    clearRoute: (state) => ({
+      ...state,
+      startLocation: null,
+      endLocation: null,
+      startAddress: '',
+      endAddress: '',
+      path: null,
+      directions: null,
+      elevationProfile: null,
+    }),
+  },
+});
+
+export const {
+  setStartAddress,
+  setEndAddress,
+  setStartLocation,
+  setEndLocation,
+  setDistance,
+  setDirections,
+  setElevationProfile,
+  setPath,
+  clearPath,
+  clearMarkers,
+  clearRoute,
+} = searchSlice.actions;
+
+export default searchSlice.reducer;
```

Comment:
Fascinating the amount of "magic" associated with `createSlice`. I haven't used this apparently more modern redux toolkit stuff before, but have used redux a bunch. It's neat how `searchSlice.actions` seems to magically exist without you defining it and how there is some sort of destinction between `searchSlice.reducer` and `searchSlice.reducers`. Am I inferring correctly that `.reducer` is a root reducer that comes from aggregating the specified `reducers`?

Reply:
Pretty much spot on. Redux-toolkit comes with a few wrapper methods like `createSlice` and `createApi` that automatically handle all the boilerplate that made Redux a little frustrating to work with in the past.

The `createApi` method is a layer on top of `createSlice` that automatically creates hooks that handle data fetching state (loading, success, error, etc) and passes them all into the component using them.

https://redux-toolkit.js.org/tutorials/rtk-query